### PR TITLE
Project SHM homework 2 and 3

### PR DESCRIPTION
**Pull request by jakjan95**
We are delivering MVP version of our SHM game to you that you can verify it and test it yourself. Game is actually working but some bugs are out of our control, we are trying to fix them but pointers are pretty weird stuff. Fell free to leave your opinion and suggestions especially. We release updates regularly to make your experience better. 😃

Authors: @jakjan95, @mariomoore, @majkel84, @MarekJanek

### CODE REVIEW

File: shm/include/fruit.hpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    }
+    std::string getName() const override { return name_; }
+    size_t getAmount() const override { return amount_; }
+    double getBasePrice() const override { return basePrice_; }
+    size_t getTimeToRotten() const { return timeToRotten_; }
+    std::string getInfo() const override;
+    friend std::ostream& operator<<(std::ostream& os, const Cargo* cargo);
+    void nextDay() override;
+
+    // Fruit& operator--();
```
Comment (nauka-programowania-MA) [inline shm/include/fruit.hpp]: remove

File: shm/include/item.hpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
+#pragma once
+#include <iostream>
+#include <map>
+//#include <string>
```
Comment (nauka-programowania-MA) [inline shm/include/item.hpp]: remove

File: shm/include/alcohol.hpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    Alcohol(const std::string& name, size_t amount, double basePrice);
+    Alcohol(const std::string& name, size_t amount, double basePrice, double volume);
+    ~Alcohol() override {}
+
+    double getPrice() const override;
+    std::string getName() const override { return name_; }
+    size_t getAmount() const override { return amount_; }
+    double getBasePrice() const override { return basePrice_; }
+    double getVolume() const { return volume_; }
+    std::string getInfo() const override;
```
Comment (nauka-programowania-MA) [inline shm/include/alcohol.hpp]: override form Cargo

File: shm/include/fruit.hpp [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+class Fruit : public Cargo {
+public:
+    Fruit(const std::string& name, size_t amount, double basePrice);
+    Fruit(const std::string& name, size_t amount, double basePrice, size_t timeToRotten);
+    ~Fruit() override {}
+
+    double getPrice() const override {
+        return getBasePrice() * static_cast<double>((getTimeToRotten()) / (getTimeToRotten() + timeElapsed_));
+    }
```
Comment (nauka-programowania-MA) [inline shm/include/fruit.hpp]: Only simply getter in .hpp. Please move to cpp

File: shm/include/item.hpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    size_t getAmount() const override { return amount_; }
+    double getBasePrice() const override { return basePrice_; }
+    Rarity getRarity() const { return rarity_; }
+    std::string getInfo() const override;
+
+    std::string enumToString(Rarity rarity) const;
+
+    void nextDay() override { 
+        // std::cout << "nextDay() for item\n";
+    }
```
Comment (nauka-programowania-MA) [inline shm/include/item.hpp]: Lack of implementation

File: shm/include/player.hpp [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+#include <memory>
+
+#include "ship.hpp"
+#include "time.hpp"
+
+class Player : public Ship::Delegate {
+public:
+    Player(std::shared_ptr<Ship> ship, size_t money);
+    Player(size_t money, std::shared_ptr<Time> time);
+    virtual ~Player() = default;
```
Comment (nauka-programowania-MA) [inline shm/include/player.hpp]: = default is definition. Please move to .cpp

File: shm/include/player.hpp [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+    std::shared_ptr<Ship> generateShip(std::shared_ptr<Time> time);
+    std::shared_ptr<Ship> getShip() const { return ship_; }
+    size_t getMoney() const { return money_; }
+    size_t getSpeed() const { return ship_->getSpeed(); }
+    cargoPtr getCargo(size_t index) const { return ship_->getCargo()[index]; }
```
Comment (nauka-programowania-MA) [inline shm/include/player.hpp]: validate index

File: shm/include/ship.hpp [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    Ship();
+    Ship(size_t id, std::string name, size_t speed, size_t maxCrew, size_t capacity);
+    Ship(size_t id, std::string name, size_t speed, size_t maxCrew, size_t capacity, Delegate* delegate, std::shared_ptr<Time> time);
+    Ship(size_t id, std::string name, size_t speed);
+    ~Ship();
+
+    int getId() const { return id_; }
+    std::string getName() const { return name_; }
+    size_t getSpeed() const { return speed_; }
+    // size_t getMaxCrew() const { return maxCrew_; }
```
Comment (nauka-programowania-MA) [inline shm/include/ship.hpp]: remove

File: shm/include/ship.hpp [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    size_t getCapacity() const { return capacity_; }
+    std::vector<cargoPtr> getCargo() const { return cargo_; }
+    size_t getCrew() const { return crew_; }
+    void load(cargoPtr cargo);
+
+    cargoPtr FindMatchCargo(Cargo* cargo);
+    Cargo* getCargo(size_t index) { return cargo_[index].get(); }
```
Comment (nauka-programowania-MA) [inline shm/include/ship.hpp]: validate index

File: shm/include/ship.hpp [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+private:
+    int id_;
+    std::string name_;
+    size_t speed_;
+    size_t maxCrew_;
+    size_t capacity_;
+    size_t crew_;
+    std::vector<cargoPtr> cargo_;
+    // size_t money_;
```
Comment (nauka-programowania-MA) [inline shm/include/ship.hpp]: delete

File: shm/include/time.hpp [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+// private:
+//     Time* time_;
+// };
+//
+// int main() {
+//     Time* time = new Time(100);
+//     ++(*time); // or
+//     (*time)++;
+//     delete time;
+// }
```
Comment (nauka-programowania-MA) [inline shm/include/time.hpp]: remove

File: shm/include/time.hpp [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+#pragma once
+
+#include <list>
+
+class Observer {
+public:
+    virtual void nextDay() = 0;
+    virtual ~Observer(){};
+};
```
Comment (nauka-programowania-MA) [inline shm/include/time.hpp]: This observer should not be standalone. Please move for example to Cargo

File: shm/include/time.hpp [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+class Time {
+public:
+    Time();
+    Time(int game_days);
+    void addObserver(Observer* obs);
+    void removeObserver(Observer* obs);
+    size_t getElapsedTime() const;
+    // Prefix increment operator (decrement elapsed time)
```
Comment (nauka-programowania-MA) [inline shm/include/time.hpp:18]: redundant

File: shm/include/time.hpp [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    Time& operator++();
+    // Postfix increment operator (decrement elapsed time)
```
Comment (nauka-programowania-MA) [inline shm/include/time.hpp:20]: redundant

File: shm/src/fruit.cpp [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+       << std::left << std::setw(5) << "\tTime to rotten: " << getTimeToRotten() << '\n';
+    return ss.str();
+}
+
+// Fruit& Fruit::operator--() {
+//     if (timeToRotten_ > 0) {
+//         timeToRotten_--;
+//     }
+//     return *this;
+// }
```
Comment (nauka-programowania-MA) [inline shm/src/fruit.cpp]: Remove

File: shm/src/cargo.cpp [hunk 16]
```diff
@@ -0,0 +1,10 @@
+#include "cargo.hpp"
+
+Cargo::Cargo(std::string name, size_t amount, double basePrice)
+    : name_(name),
+      amount_(amount),
+      basePrice_(basePrice) {}
+
+bool Cargo::operator==(Cargo& cargo) const {
+    return name_ == cargo.name_;
+}
```
Comment (nauka-programowania-MA) [inline shm/src/cargo.cpp:10]: Redundant. This could be pure virtual, because You never create instance of Cargo

File: shm/src/game.cpp [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+#include <limits>
+
+Game::Game() {}
+
+Game::Game(size_t start_money, size_t game_days, size_t final_goal)
+    : start_money_(start_money), game_days_(game_days), final_goal_(final_goal) {
+    time_ = std::make_shared<Time>(game_days_);
+    map_ = std::make_unique<Map>(time_);
+    //ship_ = generateShip();
+    //player_ = std::make_unique<Player>(ship_, start_money_);
```
Comment (nauka-programowania-MA) [inline shm/src/game.cpp]: remove

File: shm/src/game.cpp [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+        std::cin >> action;
+        makeAction(static_cast<Action>(action));
+    }
+    printLooseScreen();
+    exit();
+}
+
+// std::shared_ptr<Ship> Game::generateShip() {
+//     return std::make_shared<Ship>(1, "Maria", 1, 128, 1200);
+// }
```
Comment (nauka-programowania-MA) [inline shm/src/game.cpp]: remove

File: shm/src/game.cpp [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+    }
+    auto destination = map_->getIsland(Coordinates(x, y));
+    if (destination != nullptr) {
+        for (auto i = map_->getDistanceToIsland(destination); i > 0; --i) {
+            // one "distance" takes one day
+            ++(*time_);
+            if (time_->getElapsedTime() == 0) {
+                break;
+            }
+            // player_->nextDay();
```
Comment (nauka-programowania-MA) [inline shm/src/game.cpp:130]: not needed. Time should notify observers

File: shm/src/player.cpp [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+    return ship_->getCapacity() - totalAmount;
+}
+
+void Player::printCargo() const {
+    ship_ -> printCargo();
+}
+
+// void Player::nextDay() {
+//     ship_->nextDay();
+// }
```
Comment (nauka-programowania-MA) [inline shm/src/player.cpp]: Remove

File: shm/src/ship.cpp [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+    time_->removeObserver(this);
+}
+
+void Ship::setName(const std::string& name) {
+    name_ = name;
+}
+
+void Ship::nextDay() {
+    for (auto el : getCargo()) {
+        //*el.get() += 1;
```
Comment (nauka-programowania-MA) [inline shm/src/ship.cpp]: remove

File: shm/src/ship.cpp [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+        el.get()->nextDay();
```
Comment (nauka-programowania-MA) [inline shm/src/ship.cpp:26]: Wrong. Time should notify all observers. So register observer for Cargo. Instead, notify them when nextDay will be call for Ship.

File: shm/src/store.cpp [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+        Fruit tempFruit(cargo->getName(), amount, cargo->getBasePrice());
+        cargoPtr tempCargo = std::make_shared<Fruit>(tempFruit);
+        player->getShip()->load(tempCargo);
+    }
+    
+    if (typeid(*cargo) == typeid(Item)) {
+        Item tempItem(cargo->getName(), amount, cargo->getBasePrice());
+        cargoPtr tempCargo = std::make_shared<Item>(tempItem);
+        player->getShip()->load(tempCargo);
+    }
```
Comment (nauka-programowania-MA) [inline shm/src/store.cpp]: Here you can use also dynamic_cast. Also, you will check 3 time if. Make if else instead.

File: shm/src/time.cpp [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+// Prefix increment operator (decrement elapsed time)
+Time& Time::operator++() {
+    time_elapsed_--;
+    auto it = observers_.begin();
+    while (it != observers_.end()) {
+        (*it)->nextDay();
+        ++it;
+    }
```
Comment (nauka-programowania-MA) [inline shm/src/time.cpp]: hmm range loop will be better here.

**Review by nauka-programowania-MA — CHANGES_REQUESTED** [review]

File: shm/include/fruit.hpp (hunk 1 above)
Comment (majkel84) [reply shm/include/fruit.hpp]: done

File: shm/include/item.hpp (hunk 2 above)
Comment (majkel84) [reply shm/include/item.hpp]: done

File: shm/include/fruit.hpp (hunk 4 above)
Comment (majkel84) [reply shm/include/fruit.hpp]: done

File: shm/include/item.hpp (hunk 5 above)
Comment (majkel84) [reply shm/include/item.hpp]: nextDay don't have and implementation for Item and Alcohol, just override from interface Cargo and do nth.

File: shm/src/player.cpp (hunk 20 above)
Comment (majkel84) [reply shm/src/player.cpp]: done

File: shm/src/ship.cpp (hunk 21 above)
Comment (majkel84) [reply shm/src/ship.cpp]: done

File: shm/src/time.cpp (hunk 24 above)
Comment (majkel84) [reply shm/src/time.cpp]: done

File: shm/include/ship.hpp (hunk 8 above)
Comment (majkel84) [reply shm/include/ship.hpp]: done

File: shm/include/player.hpp (hunk 6 above)
Comment (majkel84) [reply shm/include/player.hpp]: done

File: shm/include/item.hpp [hunk 25]
```diff
@@ -0,0 +1,36 @@
+#pragma once
+#include <iostream>
+#include <map>
```
Comment (WBigaj) [inline shm/include/item.hpp:3]: I think that newline between pragma and includes would be nice

File: shm/include/store.hpp [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+    friend std::ostream& operator<<(std::ostream& out, const Store& store);
+
+private:
+    void GenerateCargo();
+    size_t GenerateRandom(size_t minValue, size_t maxValue);
+    cargoPtr FindMatchCargo(cargoPtr cargo);
+    void RemoveFromStore(cargoPtr cargo);
+    std::array<std::string, 3> fruitsNames{"apple", "raspberry", "lemon"};
+    std::array<std::string, 3> alcoNames{"vodka", "whiskey", "beer"};
+    std::array<std::string, 3> itemNames{"tools", "wood", "material"};
```
Comment (WBigaj) [inline shm/include/store.hpp]: Can't those three arrays be constexpr considering they are predetermined types of items?

**Review by WBigaj — COMMENTED** [review]
I have checked your code and given my feedback.

File: shm/src/store.cpp (hunk 23 above)
Comment (jakjan95) [reply shm/src/store.cpp]: Corrected, but i had problem with dynamic_cast so I leave it with sharded_pointer.

File: shm/include/store.hpp (hunk 26 above)
Comment (jakjan95) [reply shm/include/store.hpp]: It can't be constexpr but it can be const, thanks :)

File: shm/include/alcohol.hpp (hunk 3 above)
Comment (jakjan95) [reply shm/include/alcohol.hpp]: done

**Comment by jakjan95** [comment]
We are pending for review. We fixed all commented stuff, and some bugs. But we still have little bug in selling items method. When item is bought on current island selling works for this item too, but when we move to another island we are getting a segmentation fault during selling of that item from other island. @nauka-programowania-MA can you give a little tip how we can deal with it?

File: shm/include/ship.hpp [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+    int getId() const { return id_; }
+    std::string getName() const { return name_; }
+    size_t getSpeed() const { return speed_; }
+    size_t getCapacity() const { return capacity_; }
+    std::vector<cargoPtr> getCargo() const { return cargo_; }
+    size_t getCrew() const { return crew_; }
+    void load(cargoPtr cargo);
+
+    cargoPtr FindMatchCargo(Cargo* cargo);
+    Cargo* getCargo(const size_t index) { return cargo_[index].get(); }
```
Comment (nauka-programowania-MA) [inline shm/include/ship.hpp:33]: Are you validate somewhere index?

**Review by nauka-programowania-MA — APPROVED** [review]

**Comment by nauka-programowania-MA** [comment]
{
"requirement_ids": [2568, 2568, 2568, 2571, 2571, 2571, 2569, 2570],
"points": [5, 5, 5, 8, 8, 8, 20, 6],
"users": ["jakjan95", "mariomoore", "majkel84", "MarekJanek"]
}

**Comment by nauka-programowania-MA** [comment]
> We are pending for review. We fixed all commented stuff, and some bugs. But we still have little bug in selling items method. When item is bought on current island selling works for this item too, but when we move to another island we are getting a segmentation fault during selling of that item from other island. @nauka-programowania-MA can you give a little tip how we can deal with it?

I analyzed code. But can't find problem. I guess you delete ptr instead hold them in ship. You can easily check this using debuger and check where you exactly got crash.

**Pull request closed without merging**